internal/resources/providers/gcplib/inventory/provider.go

```diff
@@ -91,242 +75,327 @@ type dnsPolicyFields struct {
-				org, err := crmService.Organizations.Get(parent).Context(ctx).Do()
-				if err != nil {
-					log.Errorf("error fetching GCP Org: %s, error: %s", parent, err)
-					return ""
-				}
-				return org.DisplayName
-			}, parent)
-		},
+	cloudResourceManager, err := NewResourceManagerWrapper(ctx, log, gcpConfig)
+	if err != nil {
+		return nil, err
 	}
 
 	return &Provider{
-		config:                    gcpConfig,
-		log:                       log,
-		inventory:                 assetsInventoryWrapper,
-		crm:                       crmServiceWrapper,
-		cloudAccountMetadataCache: NewMapCache[*fetching.CloudAccountMetadata](),
+		config:    gcpConfig,
+		log:       log,
+		inventory: assetsInventory,
+		crm:       cloudResourceManager,
 	}, nil
 }
 
-func (p *Provider) ListAllAssetTypesByName(ctx context.Context, assetTypes []string) ([]*ExtendedGcpAsset, error) {
-	wg := sync.WaitGroup{}
-	var resourceAssets []*assetpb.Asset
-	var policyAssets []*assetpb.Asset
-
-	wg.Add(1)
-	go func() {
-		resourceAssets = p.getAllAssets(ctx, &assetpb.ListAssetsRequest{
-			Parent:      p.config.Parent,
-			AssetTypes:  assetTypes,
-			ContentType: assetpb.ContentType_RESOURCE,
-		})
-		wg.Done()
-	}()
-	wg.Add(1)
-	go func() {
-		policyAssets = p.getAllAssets(ctx, &assetpb.ListAssetsRequest{
-			Parent:      p.config.Parent,
-			AssetTypes:  assetTypes,
-			ContentType: assetpb.ContentType_IAM_POLICY,
-		})
-		wg.Done()
-	}()
+func (p *Provider) ListAssetTypes(ctx context.Context, assetTypes []string, out chan<- *ExtendedGcpAsset) {
+	defer close(out)
 
-	wg.Wait()
+	resourceCh := make(chan *assetpb.Asset) // *assetpb.Asset with Resource
+	policyCh := make(chan *assetpb.Asset)   // *assetpb.Asset with IamPolicy
+	mergeCh := make(chan *assetpb.Asset)    // *assetpb.Asset with Resource and IamPolicy
+	enrichCh := make(chan *ExtendedGcpAsset)
 
-	var assets []*assetpb.Asset
-	assets = append(append(assets, resourceAssets...), policyAssets...)
-	mergedAssets := mergeAssetContentType(assets)
-	extendedAssets := p.extendWithCloudMetadata(ctx, mergedAssets)
-	// Enrich network assets with dns policy
-	p.enrichNetworkAssets(ctx, extendedAssets)
+	go p.getAllAssets(ctx, p.config.Parent, assetpb.ContentType_RESOURCE, assetTypes, resourceCh)
+	go p.getAllAssets(ctx, p.config.Parent, assetpb.ContentType_IAM_POLICY, assetTypes, policyCh)
+	go p.mergeAssets(ctx, mergeCh, resourceCh, policyCh)
+	go p.enrichAssets(ctx, mergeCh, enrichCh)
 
-	return extendedAssets, nil
+	for asset := range enrichCh {
+		out <- asset
+	}
 }
 
-// ListMonitoringAssets returns a list of monitoring assets grouped by project id
-func (p *Provider) ListMonitoringAssets(ctx context.Context, monitoringAssetTypes map[string][]string) ([]*MonitoringAsset, error) {
-	logMetrics, err := p.ListAllAssetTypesByName(ctx, monitoringAssetTypes["LogMetric"])
-	if err != nil {
-		return nil, err
+func (p *Provider) ListMonitoringAssets(ctx context.Context, out chan<- *MonitoringAsset) {
```

Comment:
this method used to fetch all resources and policies for 2 asset types, then group them by project, and once all assets are grouped, return a slice with all groups

it now does the same thing, but sends each group once it's ready, instead of waiting for all of them to be ready. also, only resources are fetched (no policies, as CIS rules don't require it. not all these asset types necessarily have a policy, but trying to fetch them was redundant)